### PR TITLE
Port the bidaf endpoint to the new id format.

### PR DESCRIPTION
**Pull request by codeviking**
The AllenNLP team is porting model IDs to a new format, of
the form `{task}-{model}`. This PR makes changes to the
way the `bidaf` endpoint is deployed so that it uses the
new id form, which is `rc-bidaf`.

To do this in an iterative way without downtime, in this PR
I did the following:

1. Introduced `common.v1.APIEndpoint()`, which supports the
   `altPaths` argument.  The `altPaths` argument is a list of
   additional paths that will be routed to the endpoint being
   deployed. The resulting Kubernetes config sets up both
   `/api/model/:id` and `altPaths` to be directed to the
   endpoint being deployed.

2. Modified `bidaf` to use the new ID format (`rc-bidaf`) and
   the new API for configuring Kubernetes, with `altPaths` to
   to `['/api/bidaf']`. This ensures the UI code that depends
   on that path will keep working.

It's worth noting that deploying this will orphan the existing
deployment, since the ID is used to produce the name of the
Kubernetes things we deploy. This is actually nice, as if I
screw things up nothing will stop working.

If things go well and I can confirm that the new endpoint is
working after I deploy, I plan on manually cleaning up the
old environment.

Here's a diff that shows what's changed in the Kubernetes manifest:

```diff
--- prod-webapp.yml	2021-01-14 11:50:40.000000000 -0800
+++ new-webapp.yml	2021-01-14 11:58:10.000000000 -0800
@@ -7,9 +7,9 @@
          "app": "allennlp-demo",
          "contact": "allennlp-contact",
          "team": "allennlp"
       },
-      "name": "allennlp-demo-api-bidaf"
+      "name": "allennlp-demo-api-rc-bidaf"
    }
 }
 ---
 {
@@ -25,23 +25,23 @@
       },
       "labels": {
          "app": "allennlp-demo",
          "contact": "allennlp-contact",
-         "endpoint": "bidaf",
+         "endpoint": "rc-bidaf",
          "env": "prod",
          "team": "allennlp"
       },
-      "name": "allennlp-demo-api-bidaf-prod",
-      "namespace": "allennlp-demo-api-bidaf"
+      "name": "allennlp-demo-api-rc-bidaf-prod",
+      "namespace": "allennlp-demo-api-rc-bidaf"
    },
    "spec": {
       "replicas": 2,
       "revisionHistoryLimit": 3,
       "selector": {
          "matchLabels": {
             "app": "allennlp-demo",
             "contact": "allennlp-contact",
-            "endpoint": "bidaf",
+            "endpoint": "rc-bidaf",
             "env": "prod",
             "team": "allennlp"
          }
       },
@@ -55,15 +55,15 @@
             },
             "labels": {
                "app": "allennlp-demo",
                "contact": "allennlp-contact",
-               "endpoint": "bidaf",
+               "endpoint": "rc-bidaf",
                "env": "prod",
                "onlyOneOfPerNode": "allennlp-demo-prod",
                "team": "allennlp"
             },
-            "name": "allennlp-demo-api-bidaf-prod",
-            "namespace": "allennlp-demo-api-bidaf"
+            "name": "allennlp-demo-api-rc-bidaf-prod",
+            "namespace": "allennlp-demo-api-rc-bidaf"
          },
          "spec": {
             "affinity": {
                "podAntiAffinity": {
@@ -97,9 +97,9 @@
                      },
                      "initialDelaySeconds": 120,
                      "periodSeconds": 10
                   },
-                  "name": "allennlp-demo-api-bidaf-prod",
+                  "name": "allennlp-demo-api-rc-bidaf-prod",
                   "readinessProbe": {
                      "failureThreshold": 1,
                      "httpGet": {
                         "path": "/?check=readiness_probe",
@@ -133,14 +133,14 @@
       },
       "labels": {
          "app": "allennlp-demo",
          "contact": "allennlp-contact",
-         "endpoint": "bidaf",
+         "endpoint": "rc-bidaf",
          "env": "prod",
          "team": "allennlp"
       },
-      "name": "allennlp-demo-api-bidaf-prod",
-      "namespace": "allennlp-demo-api-bidaf"
+      "name": "allennlp-demo-api-rc-bidaf-prod",
+      "namespace": "allennlp-demo-api-rc-bidaf"
    },
    "spec": {
       "ports": [
          {
@@ -150,9 +150,9 @@
       ],
       "selector": {
          "app": "allennlp-demo",
          "contact": "allennlp-contact",
-         "endpoint": "bidaf",
+         "endpoint": "rc-bidaf",
          "env": "prod",
          "team": "allennlp"
       }
    }
@@ -174,14 +174,14 @@
       },
       "labels": {
          "app": "allennlp-demo",
          "contact": "allennlp-contact",
-         "endpoint": "bidaf",
+         "endpoint": "rc-bidaf",
          "env": "prod",
          "team": "allennlp"
       },
-      "name": "allennlp-demo-api-bidaf-prod",
-      "namespace": "allennlp-demo-api-bidaf"
+      "name": "allennlp-demo-api-rc-bidaf-prod",
+      "namespace": "allennlp-demo-api-rc-bidaf"
    },
    "spec": {
       "rules": [
          {
@@ -189,9 +189,16 @@
             "http": {
                "paths": [
                   {
                      "backend": {
-                        "serviceName": "allennlp-demo-api-bidaf-prod",
+                        "serviceName": "allennlp-demo-api-rc-bidaf-prod",
+                        "servicePort": 8000
+                     },
+                     "path": "/api/model/rc-bidaf(/(.*))?$"
+                  },
+                  {
+                     "backend": {
+                        "serviceName": "allennlp-demo-api-rc-bidaf-prod",
                         "servicePort": 8000
                      },
                      "path": "/api/bidaf(/(.*))?$"
```

### CODE REVIEW

File: api/allennlp_demo/common/.skiff/common.libsonnet [hunk 1]
```diff
@@ -9,8 +9,266 @@
 local config = import '../../../../skiff.json';
 local db = import 'db.libsonnet';
 
+// There are several methods for creating an APIEndpoint. This local function represents the common
+// bits that can be shared by the individual, public interfaces. Once the non `v1` interface is
+// removed this can all be inlined there.
+local APIEndpoint(id, image, cause, sha, cpu, memory, env, branch, repo, buildId, startupTime=120,
+                  useDb=false, maxBodySize = '0.5m'
+) =
```
Comment (codeviking) [inline api/allennlp_demo/common/.skiff/common.libsonnet:17]: This is all existing code, it's just a local function now that's used below. I wouldn't review this too closely.

File: api/allennlp_demo/common/.skiff/common.libsonnet [hunk 2]
```diff
@@ -38,253 +296,87 @@ local db = import 'db.libsonnet';
...
+                useDb=false, maxBodySize = '0.5m'
+    ):
+        local endpoint = APIEndpoint(id, image, cause, sha, cpu, memory, env, branch, repo, buildId,
+                                     startupTime, useDb, maxBodySize);
+        [
+            endpoint.namespace,
+            endpoint.deployment,
+            endpoint.service,
+            endpoint.ingress
+        ],
```
Comment (codeviking) [inline api/allennlp_demo/common/.skiff/common.libsonnet:309]: This is the path we're deprecating. It simply calls `APIEndpoint` and returns the results.

File: api/allennlp_demo/common/.skiff/common.libsonnet [hunk 3]
```diff
@@ -38,253 +296,87 @@ local db = import 'db.libsonnet';
...
+         *                                      Sizes can be specified in bytes, kilobytes (suffixes k
+         *                                      and K) or megabytes (suffixes m and M), for example,
+         *                                      "1024”, "8k”, "1m”.
+         *                                  NGINX docs: http://nginx.org/en/docs/syntax.html
+         * @param altPaths      {string[]}  A list of additional paths that will be routed to the model.
+         */
+        APIEndpoint(
+            id, image, cause, sha, cpu, memory, env, branch, repo, buildId, startupTime=120,
+            useDb=false, maxBodySize = '0.5m', altPaths = []
+        ):
```
Comment (codeviking) [inline api/allennlp_demo/common/.skiff/common.libsonnet:344]: This is the new interface, `v1.APIEndpoint()`. We call `APIEndpoint()` and hedge in the changes we need to make (to the ingress paths).

**Review by aimichal — COMMENTED** [review]
> This ensures the UI code that depends on that path will keep working.

Is altPaths meant to temporarily support this no-downtime migration (for bidaf and other models), or will it survive afterwards to support old client that depend on (for example) /api/bidaf?

File: api/allennlp_demo/common/.skiff/common.libsonnet [hunk 4]
```diff
@@ -38,253 +296,87 @@ local db = import 'db.libsonnet';
...
-                    metadata: {
-                        name: fullyQualifiedName,
-                        namespace: namespace.metadata.name,
-                        labels: podLabels,
-                        annotations: annotations
+            // The only difference between the `v1.APIEndpoint` instance and its former version
+            // is the `altPaths` parameter, which was added for backwards compatibility sake.
+            // Here we built up a list of paths which we'll use to override those produced
+            // by the former mechanism.
+            local pathPrefixes = [ '/api/model/' + id ] + altPaths;
```
Comment (aimichal) [inline api/allennlp_demo/common/.skiff/common.libsonnet:352]: Would it be equivalent to do something like this, without a new argument and for loop?

```js
local backend = {
  serviceName: endpoint.service.metadata.name,
  servicePort: endpoint.service.spec.ports[0].port
};

local paths = [
   { backend: backend, path: '/api/model/rc-bidaf(/(.*))?$' }
   { backend: backend, path:          '/api/bidaf(/(.*))?$' }
];
```

Comment (codeviking) [reply api/allennlp_demo/common/.skiff/common.libsonnet:352]: The `id` (`rc-bidaf` in this case) and `altPaths` are arguments.  I don't follow how I could do that via the proposed mechanism?

Comment (aimichal) [reply api/allennlp_demo/common/.skiff/common.libsonnet:352]: I thought this code was specifically in the bidaf directory, but it's in a common file.

Carry on!

**Review by aimichal — APPROVED** [review]

**Pull request merged**